marshal.c

```diff
@@ -754,6 +756,60 @@ w_objivar(VALUE obj, struct dump_call_arg *arg)
     w_ivar_each(obj, num, arg);
 }
 
+// Optimized dump for fixnum larger than 31-bits
+static void
+w_bigfixnum(VALUE obj, struct dump_arg *arg)
+{
+    RUBY_ASSERT(FIXNUM_P(obj));
+
+    w_byte(TYPE_BIGNUM, arg);
+
+#if SIZEOF_LONG == SIZEOF_VALUE
+    long num, slen_num;
+    num = FIX2LONG(obj);
+#else
+    long long num, slen_num;
+    num = NUM2LL(obj);
+#endif
+
+    char sign = num < 0 ? '-' : '+';
+    w_byte(sign, arg);
+
+    // Guaranteed not to overflow, as FIXNUM is 1-bit less than long
+    if (num < 0) num = -num;
+
+    // calculate the size in shorts
+    int slen = 0;
+    {
+        slen_num = num;
+        while (slen_num) {
+            slen++;
+            slen_num = SHORTDN(slen_num);
+        }
+    }
+
+    RUBY_ASSERT(slen > 0 && slen <= SIZEOF_LONG / 2);
+
+    w_long((long)slen, arg);
+
+    for (int i = 0; i < slen; i++) {
+        w_short(num & SHORTMASK, arg);
+        num = SHORTDN(num);
+    }
+
+    // We aren't adding this object to the link table, but we need to increment
+    // the index.
+    arg->num_entries++;
```

Comment:
Oh so in 3.1 it was added to `st_add_direct(arg->data` but because the Bignum was generated dynamically it was everytime a new one, hence it feels an unintentional bug of then to not use links for "big fixnums": https://github.com/ruby/ruby/blob/ruby_3_1/marshal.c#L800

Reply:
Yeah, as you say this didn't change behaviour. As far back as I checked (not very far, just 2.7) Ruby didn't de-dup "large fixnums".

In the commit message:
> This commit avoids allocating the intermediate bignum object, instead
> outputting the T_FIXNUM directly to a Marshal bignum. This maintains the
> same representation as the previous implementation, including not using
> LINKs for these large fixnums (an artifact of the previous
> implementation always allocating a new BIGNUM).

I don't know whether or not the non-linking was intentional, I can see the argument for it being a bug but IMO either representation is valid, as the only case this round trips differently is when transferring between a 32-bit and 64-bit Ruby and those can't have the same in-memory representation anyways.

Both because it's faster (we skip not only the expensive st_add but also an st_lookup) and because it preserved existing behaviour I went with this approach.

Are you seeing any compatibility issues with this change, or are you just looking to pin down behaviour?

Reply:
Thank you for the reply, yeah I guess it was done to keep the exact same marshal output as with 32-bits Ruby where there are no "large fixnums". We're fixing a bug related to this in TruffleRuby and were wondering why the behavior is so complicated and rather inconsistent there.